core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala

```diff
@@ -118,8 +125,21 @@ class OpenHashSet[@specialized(Long, Int, Double, Float) T: ClassTag](
    * See: https://issues.apache.org/jira/browse/SPARK-45599
    */
   @annotation.nowarn("cat=other-non-cooperative-equals")
-  private def keyExistsAtPos(k: T, pos: Int) =
-    _data(pos) equals k
+  protected def keyExistsAtPos(k: T, pos: Int) = {
+    classTag[T] match {
+      case ClassTag.Long => _data(pos) equals k
+      case ClassTag.Int => _data(pos) equals k
+      case ClassTag.Double => _data(pos) equals k
+      case ClassTag.Float => _data(pos) equals k
+      case _ => nonClassTagKeyExistsAtPos(k, _data(pos))
+
+    }
+  }
+
+  @annotation.nowarn("cat=other-non-cooperative-equals")
+  protected def nonClassTagKeyExistsAtPos(k: T, dataAtPos: T): Boolean = {
+    dataAtPos equals k
+  }
```

Comment:
related to the comment above, could you explain what this does?

Reply:
@uros-db please see comment above and let me know if we are on the same page